### ISSUE TITLE
`FluxGroupBy`'s `GroupedFlux` does not propagate the `Subscription` to the second (rejected) `Subscriber`

### ISSUE THREAD

**Issue by AramMessdaghi9001** [reactor/reactor-core#3554]
When calling flatMap on a Publisher which throws an exception upon subscription, there are certain cases where checkTerminated throws a NPE and masks the actual exception. Furthermore flatMap never terminates.

## Expected Behavior
flatMap terminates with the actual Error

## Actual Behavior
NPE and no Termination. Stacktrace from the example below

```
2023-08-02 13:04:02  ERROR Operators - Operator called default onErrorDropped
java.lang.NullPointerException: null
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:840) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:609) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:589) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onError(FluxFlatMap.java:452) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxGroupBy$UnicastGroupedFlux.subscribe(FluxGroupBy.java:721) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:62) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxRetry$RetrySubscriber.resubscribe(FluxRetry.java:117) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxRetry.subscribeOrReturn(FluxRetry.java:52) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.Flux.subscribe(Flux.java:8759) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:427) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxGroupBy$GroupByMain.drainLoop(FluxGroupBy.java:393) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxGroupBy$GroupByMain.drain(FluxGroupBy.java:329) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxGroupBy$GroupByMain.onNext(FluxGroupBy.java:207) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxArray$ArraySubscription.slowPath(FluxArray.java:127) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxArray$ArraySubscription.request(FluxArray.java:100) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxGroupBy$GroupByMain.onSubscribe(FluxGroupBy.java:171) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxArray.subscribe(FluxArray.java:53) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.FluxArray.subscribe(FluxArray.java:59) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.core.publisher.Flux.subscribe(Flux.java:8773) ~[reactor-core-3.5.8.jar:3.5.8]
	at reactor.test.DefaultStepVerifierBuilder$DefaultStepVerifier.toVerifierAndSubscribe(DefaultStepVerifierBuilder.java:891) ~[reactor-test-3.5.8.jar:3.5.8]
	at reactor.test.DefaultStepVerifierBuilder$DefaultStepVerifier.verify(DefaultStepVerifierBuilder.java:831) ~[reactor-test-3.5.8.jar:3.5.8]
	at reactor.test.DefaultStepVerifierBuilder$DefaultStepVerifier.verify(DefaultStepVerifierBuilder.java:823) ~[reactor-test-3.5.8.jar:3.5.8]
	at reactor.test.DefaultStepVerifierBuilder.verifyErrorSatisfies(DefaultStepVerifierBuilder.java:685) ~[reactor-test-3.5.8.jar:3.5.8]
```
## Steps to Reproduce
```kotlin
    @Test
    fun flatMapExceptionTest() {
        StepVerifier.create(Flux.just(1, 2, 3, 4, 5, 6)
            .groupBy {
                it > 3
            }
            .flatMap {
                it.flatMap<Int> { Mono.error(Error("whoopsies")) }.retry(2)
            }
        ).verifyErrorSatisfies {
            assertThat(it).hasMessage("whoopsies")
        }
    }
```
The actual error here (GroupedFlux allows only one Subscriber) is hidden by the NPE

## Possible Solution
Null check? Not sure if there is more behind it.

## Your Environment
* Reactor version(s) used: 3.5.8
* JVM version (`java -version`): openjdk version "11.0.17" 2022-10-18 LTS
* OS and version (eg `uname -a`): Linux S0155454 6.2.0-26-generic

**Comment by OlegDokuka** [comment]
this is a partial bug in FluxGroupBy, however, just a reminder that give GroupedFlux can be subscribed only once

**Comment by OlegDokuka** [comment]
Fixed in #3555
